### ISSUE TITLE
Runtime error :  missing keys in state_dict. 6 missing keys when strict =True in model.load_state_dict(csd, strict=True)

### ISSUE THREAD

**Issue by 920703** [ultralytics/yolov5#6759]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am training PASCAL VOC dataset using yolov5 model in Docker. I have run the command 
**python3 train.py --data VOC.yaml --epochs 50 --batch-size 4 --weights yolov5x.pt**

There is a line below model summary:- 
**Transferred 739/745 items from yolov5x.pt**  See below:
![123](https://user-images.githubusercontent.com/100197149/155464717-939ec049-f687-45dd-990b-68f358de670a.jpg)

I'm not able to understand why the all items are not transferred. Why not 745/745?
Usually when i train, sometimes i get 745/745 and sometimes 739/745 items.

So i have made changes in train.py file and set the **model.load_state_dict(csd, strict=True)**
now it gives runtime error.It is showing 6 missing keys. see below:
![image](https://user-images.githubusercontent.com/100197149/155465392-bdb14ed2-c6f9-4bea-b604-6601d4da12b6.png)

Although this error can be set back after changing strict= False.

But what is the reason behind this? What changes need to be done so that it always show all the items i.e 745/745.. not any missing items i.e 739/745
secondly, if it shows 739/745. It runs successfully, What effect will be there on results?
Can anyone please help me with the solution?

**Comment by glenn-jocher** [comment]
@920703 all layers with matching names and shapes are transferred.

**Comment by 920703** [comment]
@glenn-jocher but what happens if all the matching names and shapes are not transferred i.e if it shows 739/745. 
Although i get to run it successfully.
Does it make any differences in result with 745/745 and 739/745?

**Comment by glenn-jocher** [comment]
@920703 if you train a VOC dataset with 20 classes starting from a pretrained COCO model with 80 classes why would you think that every layer in YOLOv5 would have the same shape?

**Comment by 920703** [comment]
okay. but sometimes why it shows 745/745, that means at that time all the pairs are transferred.

**what i understand is that, it depends on number of classes in dataset using with pre-trained model.** 
but again, if it depends on number of classes then why everytime training voc dataset (20 classes), why sometimes it shows up with 745//745.

**Comment by glenn-jocher** [comment]
@920703 as I said all layers with matching shapes and names are transferred, i.e. train COCO from a pretrained COCO model (or any other 80 class model) and all layers will transfer.

**Comment by 920703** [comment]
Okay thank you so much. Can you clear one more doubt, please.

After training VOC dataset, i want to check accuracy on test dataset. i used the following command:
python3 detect.py --source /datasets/VOC/images/test2007 --img 512 --data data/VOC.yaml --weights runs/train/exp60/weights/best.pt

It only saves the detection results of pictures.  
How to check accuracy?

Also it is not detecting results of some pictures( around 11 or 12 pictures out of 4952 test data)
![resultsnotdetecting](https://user-images.githubusercontent.com/100197149/155831467-ce4597cf-3d75-4fe3-85f4-e7bef789a5ab.png)

i used the same test data as the validation data, as mentioned in VOC.yaml file.

**Comment by glenn-jocher** [comment]
@920703 you want to run val.py on your trained weights:
```
python val.py --data VOC.yaml --weights path/to/best.pt --img 512
```

**Comment by 920703** [comment]
@glenn-jocher I have organized my data in VOC file as: 
taken only 2012 dataset
in train = only train 2012 images
in val = only val 2012 
in test = only test2012 

When we run train.py file,  after every epoch it reports the metrics for validation data. and also at the end of training it report metrics for all the dataset classes. These metrics are for validation images.
Metrics are: 
all classes :-    P= 0.799 , R= 0.757 , map@0.5= 0.817 , mao@0.5:0.95= 0.645

For validation images i got
all classes :-       P= 0.787 , R= 0.768 , map@0.5= 0.817 , mao@0.5:0.95= 0.645

Now i run val.py with argument  as task as test

For test i got
![testmetrics](https://user-images.githubusercontent.com/100197149/156041447-0ea6f779-e798-42d6-b499-891d1749580c.png)

Why the precision and mAP is so less for testing data.
Is my data overfitted?
